drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright (c) 2020, Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT quicklogic_eos_s3_gpio
+
+#include <errno.h>
+#include <drivers/gpio.h>
+#include <soc.h>
+#include <eoss3_hal_gpio.h>
+#include <eoss3_hal_pads.h>
+#include <eoss3_hal_pad_config.h>
+
+#include "gpio_utils.h"
+
+#define MAX_GPIOS		8U
+#define GPIOS_MASK		(BIT(MAX_GPIOS) - 1)
+
+struct gpio_eos_s3_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	/* Pin configuration to determine whether use primary
+	 * or secondary pin for a target GPIO. Secondary pin is used
+	 * when the proper bit is set to 1.
+	 *
+	 * bit_index : primary_pin_number / secondary_pin_number
+	 *
+	 * 0 : 6 / 24
+	 * 1 : 9 / 26
+	 * 2 : 11 / 28
+	 * 3 : 14 / 30
+	 * 4 : 18 / 31
+	 * 5 : 21 / 36
+	 * 6 : 22 / 38
+	 * 7 : 23 / 45
+	 */
+	uint8_t pin_secondary_config;
+};
+
+struct gpio_eos_s3_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	/* port ISR callback routine address */
+	sys_slist_t callbacks;
+	/* runtime configuration */
+	uint8_t gpio_irq_num[MAX_GPIOS];
+};
+
+/* connection table to configure GPIOs with pads */
+static const PadConfig pad_configs[] = {
+	{.ucPin = PAD_6, .ucFunc = PAD6_FUNC_SEL_GPIO_0},
+	{.ucPin = PAD_9, .ucFunc = PAD9_FUNC_SEL_GPIO_1},
+	{.ucPin = PAD_11, .ucFunc = PAD11_FUNC_SEL_GPIO_2},
+	{.ucPin = PAD_14, .ucFunc = PAD14_FUNC_SEL_GPIO_3},
+	{.ucPin = PAD_18, .ucFunc = PAD18_FUNC_SEL_GPIO_4},
+	{.ucPin = PAD_21, .ucFunc = PAD21_FUNC_SEL_GPIO_5},
+	{.ucPin = PAD_22, .ucFunc = PAD22_FUNC_SEL_GPIO_6},
+	{.ucPin = PAD_23, .ucFunc = PAD23_FUNC_SEL_GPIO_7},
+	{.ucPin = PAD_24, .ucFunc = PAD24_FUNC_SEL_GPIO_0},
+	{.ucPin = PAD_26, .ucFunc = PAD26_FUNC_SEL_GPIO_1},
+	{.ucPin = PAD_28, .ucFunc = PAD28_FUNC_SEL_GPIO_2},
+	{.ucPin = PAD_30, .ucFunc = PAD30_FUNC_SEL_GPIO_3},
+	{.ucPin = PAD_31, .ucFunc = PAD31_FUNC_SEL_GPIO_4},
+	{.ucPin = PAD_36, .ucFunc = PAD36_FUNC_SEL_GPIO_5},
+	{.ucPin = PAD_38, .ucFunc = PAD38_FUNC_SEL_GPIO_6},
+	{.ucPin = PAD_45, .ucFunc = PAD45_FUNC_SEL_GPIO_7},
+};
+
+static PadConfig gpio_eos_s3_pad_select(const struct device *dev,
```

Comment:
Sorry, I wasn't very precise in my last request. It's a minor improvement since PadConfig is small, however in the code 
```
gpio_cfg.xPadConf = &pad_config;
```
we need the address of the PadConfig and not the structure itself. We could return here a pointer to PadConfig instead.

Reply:
I have warned about const-ness but didn't notice there is no way around it. What I have asked for was a bad idea. You removed const from
```
static const PadConfig pad_configs[] = {
```
and this moved the array to RAM. `PadConfig` is actually not that small, sizeof(PadConfig) is 16. So the the whole array takes 256 bytes. That's a lot of RAM. Even for flash it's not negligible.

Let's either go to the previous design or we could optimize this code. All `PAD*_FUNC_SEL_GPIO_*` have identical value. It's unlikely that in the future this will change. If you add a compile time assert using `BUILD_ASSERT` that would verify that all `PAD*_FUNC_SEL_GPIO_*` are equal then you could
```
#define PAD_FUNC_SEL_GPIO PAD6_FUNC_SEL_GPIO_0
```
and use `PAD_FUNC_SEL_GPIO` in the code.

Then we would only need an array to store relation between gpio_num and pad number.

This `BUILD_ASSERT` is going to look ugly. If you don't like it let's go to the previous design. Sorry about all that.

Reply:
No problem, I reverted changes and went back to the previous implementation.